Panic when reading transit ed25519 key when audit file backend is enabled

Summary:

When I try to read the details of a key created in the transit secret backend and the audit file backend is enabled, I get a panic.

Details:

* Vault v0.7.3
* Consul v0.8.5
* macOS 10.12.5

I am setting up a test Vault with consul storage. I start with an empty consul, just bootstrapped, initialise the vault, unseal it, mount the transit backend, create the key with type ed25519 and try to read it. This returns the details of the key. I then enable the audit file backend, try to read the key and I get a panic.

Configuration:

vault.hcl
```
storage "consul" {
  address = "localhost:8500"
  path    = "vault/"
}

listener "tcp" {
  address     = "0.0.0.0:8200"
  tls_disable = 1
}
```

consul.json
```
{ 
  "http_config": {
    "response_headers": {
      "Access-Control-Allow-Origin": "*"
    }
  },
  "client_addr": "127.0.0.1",
  "datacenter": "dc1",
  "acl_datacenter": "dc1",
  "acl_master_token": "<uuidgen token>",
  "data_dir": "path-to-consul-data",
  "server": true,
  "bootstrap": true,
  "ui": true
}
```

Steps:

1) start `consul` with `consul agent -config-file=./consul/consul.json`
2) start `vault` with `vault server -config=./vault/vault.hcl`
3) init vault, unseal it
4) `vault mount transit`
5) create key with `curl -H 'X-Vault-Token: <root token>' -X POST http://localhost:8200/v1/transit/keys/boom -d '{ "type": "ed25519" }'`
6) view the key 
```
$ curl -H 'X-Vault-Token: <root token>' -X GET http://localhost:8200/v1/transit/keys/boom
{"request_id":"901cc0d1-32f5-a98b-63c6-1792de1f6b91","lease_id":"","renewable":false,"lease_duration":0,"data":{"deletion_allowed":false,"derived":false,"exportable":false,"keys":{"1":{"name":"ed25519","public_key":"Vd1fvzQgF9RKzsWkxNguR4Wz04mrQ2yy7q1CbnVlZ4I=","creation_time":"2017-07-02T19:27:34.983833598-03:00"}},"latest_version":1,"min_decryption_version":1,"name":"boom","supports_decryption":false,"supports_derivation":true,"supports_encryption":false,"supports_signing":true,"type":"ed25519"},"wrap_info":null,"warnings":null,"auth":null}
```
7) enable the audit backend
```
$ vault audit-enable file file_path=/Users/bruno/code/evergiving-vault/vault/logs/vault_audit.log
Successfully enabled audit backend 'file' with path 'file'!
```
8) try to read the key again
```
$ curl -H 'X-Vault-Token: <root token>' -X GET http://localhost:8200/v1/transit/keys/boom
{"errors":["internal error"]}
```
9) the logs on the console from vault:
```
2017/07/02 19:25:12.183020 [INFO ] core/startClusterListener: serving cluster requests: cluster_listen_address=[::]:8201
2017/07/02 19:26:23.172309 [INFO ] core: successful mount: path=transit/ type=transit
2017/07/02 19:28:13.366782 [INFO ] core: enabled audit backend: path=file/ type=file
2017/07/02 19:28:15.129826 [ERROR] audit: panic during logging: request_path=transit/keys/boom error="reflect: reflect.Value.Set using unaddressable value"
2017/07/02 19:28:15.129852 [ERROR] core: failed to audit response: request_path=transit/keys/boom error=1 error occurred:

* panic generating audit log
```
10) the audit log only shows the request 
```
{"time":"2017-07-02T22:42:01Z","type":"request","auth":{"client_token":"hmac-sha256:0dd264d0f3158ac0af34ff0a7bac620c5a221c0d6dfe051d41e295d918a93212","accessor":"hmac-sha256:2a738dd5d8e2006bc36e0b3594ab6e22c8899033880ce6460d5661d56c7312ad","display_name":"root","policies":["root"],"metadata":null},"request":{"id":"94857778-32e6-3594-b1e1-9257c623ce3b","operation":"read","client_token":"hmac-sha256:0dd264d0f3158ac0af34ff0a7bac620c5a221c0d6dfe051d41e295d918a93212","client_token_accessor":"hmac-sha256:2a738dd5d8e2006bc36e0b3594ab6e22c8899033880ce6460d5661d56c7312ad","path":"transit/keys/boom","data":null,"remote_address":"127.0.0.1","wrap_ttl":0,"headers":{}},"error":""}
```

--- a/builtin/logical/transit/path_keys.go
+++ b/builtin/logical/transit/path_keys.go
@@ -9,6 +9,7 @@ import (
 
 	"golang.org/x/crypto/ed25519"
 
+	"github.com/fatih/structs"
 	"github.com/hashicorp/vault/helper/keysutil"
 	"github.com/hashicorp/vault/logical"
 	"github.com/hashicorp/vault/logical/framework"
@@ -155,9 +156,9 @@ func (b *backend) pathPolicyWrite(
 
 // Built-in helper type for returning asymmetric keys
 type asymKey struct {
-	Name         string    `json:"name"`
-	PublicKey    string    `json:"public_key"`
-	CreationTime time.Time `json:"creation_time"`
+	Name         string    `json:"name" structs:"name"`
+	PublicKey    string    `json:"public_key" structs:"public_key"`
+	CreationTime time.Time `json:"creation_time" structs:"creation_time"`
 }
 
 func (b *backend) pathPolicyRead(
@@ -225,7 +226,7 @@ func (b *backend) pathPolicyRead(
 		resp.Data["keys"] = retKeys
 
 	case keysutil.KeyType_ECDSA_P256, keysutil.KeyType_ED25519:
-		retKeys := map[string]asymKey{}
+		retKeys := map[string]map[string]interface{}{}
 		for k, v := range p.Keys {
 			key := asymKey{
 				PublicKey:    v.FormattedPublicKey,
@@ -254,7 +255,7 @@ func (b *backend) pathPolicyRead(
 				key.Name = "ed25519"
 			}
 
-			retKeys[strconv.Itoa(k)] = key
+			retKeys[strconv.Itoa(k)] = structs.New(key).Map()
 		}
 		resp.Data["keys"] = retKeys
 	}

--- a/builtin/logical/transit/path_keys.go
+++ b/builtin/logical/transit/path_keys.go
@@ -156,9 +156,9 @@ func (b *backend) pathPolicyWrite(
 
 // Built-in helper type for returning asymmetric keys
 type asymKey struct {
-	Name         string    `json:"name" structs:"name"`
-	PublicKey    string    `json:"public_key" structs:"public_key"`
-	CreationTime time.Time `json:"creation_time" structs:"creation_time"`
+	Name         string    `json:"name" structs:"name" mapstructure:"name"`
+	PublicKey    string    `json:"public_key" structs:"public_key" mapstructure:"public_key"`
+	CreationTime time.Time `json:"creation_time" structs:"creation_time" mapstructure:"creation_time"`
 }
 
 func (b *backend) pathPolicyRead(

--- a/builtin/logical/transit/path_keys_test.go
+++ b/builtin/logical/transit/path_keys_test.go
@@ -1,0 +1,79 @@
+package transit_test
+
+import (
+	"testing"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/audit"
+	"github.com/hashicorp/vault/builtin/audit/file"
+	"github.com/hashicorp/vault/builtin/logical/transit"
+	vaulthttp "github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/vault"
+)
+
+func TestTransit_Issue_2958(t *testing.T) {
+	coreConfig := &vault.CoreConfig{
+		LogicalBackends: map[string]logical.Factory{
+			"transit": transit.Factory,
+		},
+		AuditBackends: map[string]audit.Factory{
+			"file": file.Factory,
+		},
+	}
+
+	cluster := vault.NewTestCluster(t, coreConfig, true)
+	cluster.StartListeners()
+	defer cluster.CloseListeners()
+
+	cores := cluster.Cores
+
+	cores[0].Handler.Handle("/", vaulthttp.Handler(cores[0].Core))
+	cores[1].Handler.Handle("/", vaulthttp.Handler(cores[1].Core))
+	cores[2].Handler.Handle("/", vaulthttp.Handler(cores[2].Core))
+
+	vault.TestWaitActive(t, cores[0].Core)
+
+	client := cores[0].Client
+
+	err := client.Sys().EnableAuditWithOptions("file", &api.EnableAuditOptions{
+		Type: "file",
+		Options: map[string]string{
+			"file_path": "/dev/null",
+		},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = client.Sys().Mount("transit", &api.MountInput{
+		Type: "transit",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Logical().Write("transit/keys/foo", map[string]interface{}{
+		"type": "ecdsa-p256",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Logical().Write("transit/keys/bar", map[string]interface{}{
+		"type": "ed25519",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Logical().Read("transit/keys/foo")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Logical().Read("transit/keys/bar")
+	if err != nil {
+		t.Fatal(err)
+	}
+}

--- a/builtin/logical/transit/path_sign_verify_test.go
+++ b/builtin/logical/transit/path_sign_verify_test.go
@@ -8,6 +8,7 @@ import (
 	"golang.org/x/crypto/ed25519"
 
 	"github.com/hashicorp/vault/logical"
+	"github.com/mitchellh/mapstructure"
 )
 
 func TestTransit_SignVerify_P256(t *testing.T) {
@@ -331,8 +332,12 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
 			if err != nil {
 				t.Fatal(err)
 			}
-			val := keyReadResp.Data["keys"].(map[string]asymKey)[strings.TrimPrefix(splitSig[1], "v")]
-			if val.PublicKey != "" {
+			val := keyReadResp.Data["keys"].(map[string]map[string]interface{})[strings.TrimPrefix(splitSig[1], "v")]
+			var ak asymKey
+			if err := mapstructure.Decode(val, &ak); err != nil {
+				t.Fatal(err)
+			}
+			if ak.PublicKey != "" {
 				t.Fatal("got non-empty public key")
 			}
 			keyReadReq.Data = map[string]interface{}{
@@ -342,9 +347,12 @@ func TestTransit_SignVerify_ED25519(t *testing.T) {
 			if err != nil {
 				t.Fatal(err)
 			}
-			val = keyReadResp.Data["keys"].(map[string]asymKey)[strings.TrimPrefix(splitSig[1], "v")]
-			if val.PublicKey != base64.StdEncoding.EncodeToString(pubKeyRaw.([]byte)) {
-				t.Fatalf("got incorrect public key; got %q, expected %q", val.PublicKey, pubKeyRaw)
+			val = keyReadResp.Data["keys"].(map[string]map[string]interface{})[strings.TrimPrefix(splitSig[1], "v")]
+			if err := mapstructure.Decode(val, &ak); err != nil {
+				t.Fatal(err)
+			}
+			if ak.PublicKey != base64.StdEncoding.EncodeToString(pubKeyRaw.([]byte)) {
+				t.Fatalf("got incorrect public key; got %q, expected %q\nasymKey struct is\n%#v", ak.PublicKey, pubKeyRaw, ak)
 			}
 		}
 	}

